### ISSUE TITLE
Generated python code by antlr4 can't handle many `else if`.

### ISSUE THREAD

**Issue by jiamo** [antlr/antlr4#984]
Follow this https://github.com/jiamo/anltr4-python-bug-demo
The C grammer is clone from https://github.com/antlr/grammars-v4.
`make C`
`python3 ctest.py test_bug_c.c `

We can get error like this 

168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
    File "/usr/local/lib/python3.4/dist-packages/antlr4/PredictionContext.py", line 168, in __str__
      up = "" if self.parentCtx is None else str(self.parentCtx)
  RuntimeError: maximum recursion depth exceeded while calling a Python object

I think there is a bug for the generated C parsed code which can't handle the two many `else if` without `else` statment.

If I reduce the number of `else if` , it will pass test.

**Comment by ericvergnaud** [comment]
Hi,
thanks for this.
as mentioned by Python, the error is: maximum recursion depth exceeded
this is not a bug in the generated code, rather a limitation of Python
a simple way to solve the issue here is to change the maximum recursion depth using sys.setrecursionlimit()

**Comment by jiamo** [comment]
I am not sure, this method to solve the problem is beatiful.  The test.c on the syntax can be accept by the grammer,  so the python generated code should not met this problem in general.

Any way thanks your comment. I will have a try.

**Comment by jiamo** [comment]
Ok, sys.setrecursionlimit(2000) can make the test.c pass.  However  I am not sure what c file will be handled by my program.

**Comment by ericvergnaud** [comment]
Good to hear that the issue is solved.
Please close it. ANTLR is not meant to work around language limitations.

Envoyé de mon iPhone

**Comment by jimidle** [comment]
Hmmm - I am not sure that I 100% agree with this statement - more like
"only if there is no reasonably practical way to do so" in my book :)

However, I wonder if we still have the option to cause the generation of
switch statements rather than serial if then else? This is a huge
performance improvement for the v3 C target - so much so that I changed the
defaults to always do that unless told not to.

Jim

**Comment by ericvergnaud** [comment]
Hi Jim,
the serial 'else if' are in the parsed C sample, not in the generated code.
this results in a deep recursion loop in the runtime.
I believe there is no reasonable practical way to run the antlr runtime without recursion.
Eric

**Comment by jimidle** [comment]
Ah - I see. I was referring to the generated Python code of course. This
makes a lot more sense now :)

**Comment by jiamo** [comment]
Well, I think ply + pycparser can handle it without modify setrecursionlimit.  After set the setrecursionlimit size bigger my parse program will run longger.  So I think there may be  some issues in this situation. Can somebody explain the difference between ply and anltr4 generated parsed code?

**Comment by jiamo** [comment]
I add a example to show the performance.  https://github.com/jiamo/anltr4-python-bug-demo/blob/master/c-to-c.py. Using the simple test_bug_c.c.  we can see the performace : 

    anltr4-python-bug-demo git:(master) ✗ time python c-to-c.py test_bug_c.c
    python c-to-c.py test_bug_c.c  0.08s user 0.03s system 90% cpu 0.120 total
    anltr4-python-bug-demo git:(master) ✗ time python3 ctest.py test_bug_c.c
    ANTLR runtime and generated code versions disagree: 4.4.1!=4.5.1
    ANTLR runtime and generated code versions disagree: 4.4.1!=4.5.1
    python3 ctest.py test_bug_c.c  3.44s user 0.03s system 98% cpu 3.514 total

**Comment by ericvergnaud** [comment]
compared to other parser generators, antlr is very flexible: it makes it very easy to alter the grammar, navigate the parse tree, manage errors, insert missing tokens or skip extraneous ones. It supports semantic predicates (does ply do that?)
and antlr supports multiple target languages
all this comes at a cost, which with Python is significant - not saying the performance cannot be improved
notably, antlr has a warmup phase, so you need to run your test at least 2 times in sequence and measure the performance of the second run

**Comment by jiamo** [comment]
@ericvergnaud  I totally agree with you.  Also when I add more `else if` I should add the `setrecursionlimit` too to make pass. https://github.com/jiamo/antlr4-python-bug-demo/blob/master/ctest.py. I must set to `4000` to make it pass. I mean it may have some issues in generated code to handle this situation (which is not the python limit).

    ➜  anltr4-python-bug-demo git:(master) ✗ time python3 ctest.py test_bug_c.c
    python3 ctest.py test_bug_c.c  18.85s user 0.17s system 97% cpu 19.531 total
    ➜  anltr4-python-bug-demo git:(master) ✗ time python3 ctest.py test_bug_c.c
    python3 ctest.py test_bug_c.c  18.43s user 0.13s system 98% cpu 18.792 total
    ➜  anltr4-python-bug-demo git:(master) ✗ time python c-to-c.py test_bug_c.c
    python c-to-c.py test_bug_c.c  0.09s user 0.03s system 95% cpu 0.123 total
    ➜  anltr4-python-bug-demo git:(master) ✗

As you have said `I believe there is no reasonable practical way to run the antlr runtime without recursion.` can you give some explains or implement details on this recursion?

**Comment by ericvergnaud** [comment]
2 times in sequence means in the same process i.e. your performance test should look like:

run_test() #warmup
reset_counters()
run_test();
show_counters()

**Comment by ericvergnaud** [comment]
The grammar rules being parsed are:
```
statement:
    compoundStatement
    | selectionStatement
    ...

selectionStatement:
    'if' '(' expression ')' statement ('else' statement)?
    ...
```
What antlr does when encountering your input is as follows:
- match 'if'
- match '('
- match expression rule
- match ')'
- match statement rule (this is not recursing because { ; } is a compoundStatement, not a selectionStatement)
- match 'else' successfully
- match statement rule (this is recursive because when encountering 'if', statement will fall into selectionStatement)

So basically, each 'else if' calls the statement rule parsing code recursively. And that code computes a now very deep hierarchy of PredictionContext.

Since the grammar is itself recursive (a major antlr4 feature), the generated parser is also almost necessarily recursive.

**Comment by sharwell** [comment]
You can use `ParserInterpreter` to avoid a major source of recursion (the generated code), but the implementation of the closure operation within prediction will still be recursive. However, you would likely see a reduction in performance to use this approach.

**Comment by ericvergnaud** [comment]
Actually, thanks to your very slow test case, I was finally able to validate some performance improvements which had been pending for some time because they actually very slightly slow down small inputs.
With your test case, the performance boost can go up to 550%!
You can get the corresponding code here: https://github.com/antlr/antlr4/pull/987

**Comment by jimidle** [comment]
On Tue, Aug 25, 2015 at 10:53 PM, Jia Mo <notifications@github.com> wrote:

> As you have said I believe there is no reasonable practical way to run
> the antlr runtime without recursion. can you give some explains or
> implement details on this recursion?
>
ANTLR generates code that is more or less human readable and can be
debugged. One of the elements of this is that a parser rule is generated as
a function or function equivalent in the target language. This means that
with so many else if statements, you end up with a very deep recursion
stack. However it seems that so many else if statements are not likely in
the real world other than perhaps a the output of generated code or
artificial tests.

The recursion occurs because:

statement: if | ... | ... ;
if: 'if' expression statement 'else' statement ;

Generates a function for statement and a function for if. The parser then
becomes:

statement()
if()
statement()
if()
....

For as many else if statements as you have in the source code to be parsed.

There are of course other ways to write a parser, but this is the way ANTLR
works and in general there is no issue with the recursion depth.

Jim
